experimental/oak_async/tests/tests.rs

```diff
@@ -0,0 +1,281 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+mod dummy_data;
+mod fake_runtime;
+
+use dummy_data::DummyData;
+use fake_runtime::{add_ready_data, set_error, set_wait_on_channels_handler};
+use log::LevelFilter;
+use oak::{
+    io::{Decodable, Receiver},
+    ChannelReadStatus, OakError, OakStatus, ReadHandle,
+};
+use oak_abi::Handle;
+use oak_async::{block_on, ChannelRead, ReceiverAsync};
+
+fn init() {
+    let _ = env_logger::builder()
+        .is_test(true)
+        .filter_level(LevelFilter::Trace)
+        .try_init();
+}
+
+#[test]
+fn block_on_resolve_immediate() {
+    init();
+    let result = block_on(async { 42 }).expect("block_on failed");
+
+    assert_eq!(result, 42);
+}
+
+#[test]
+fn block_on_channel_read_immediate_ready() {
+    init();
+    add_ready_data(1, &DummyData::new("data"));
+    let read_result: Result<DummyData, OakError> =
+        block_on(do_channel_read(1)).expect("block_on failed");
+
+    assert_eq!(
+        read_result.expect("Read returned error"),
+        DummyData::new("data")
+    )
+}
+
+#[test]
+fn block_on_channel_read_ready_after_wait() {
+    init();
+    set_wait_on_channels_handler(|handles| {
+        assert_eq!(handles.len(), 1);
+        assert_eq!(handles[0].handle(), 123);
+        add_ready_data(123, &DummyData::new("data"));
+        handles[0].set_status(ChannelReadStatus::ReadReady);
+        OakStatus::Ok
+    });
+
+    let read_result: Result<DummyData, OakError> =
+        block_on(do_channel_read(123)).expect("block_on failed");
+
+    assert_eq!(
+        read_result.expect("Read returned error"),
+        DummyData::new("data")
+    )
+}
+
+#[test]
+fn block_on_multiple_readers_sequentially() {
+    init();
+    set_wait_on_channels_handler(|handles| {
+        assert_eq!(handles.len(), 1);
+        let data = match handles[0].handle() {
+            123 => "Hello",
+            456 => "world!",
+            c => panic!("Unexpected waiting channel {}", c),
+        };
+        add_ready_data(handles[0].handle(), &DummyData::new(data));
+        handles[0].set_status(ChannelReadStatus::ReadReady);
+        OakStatus::Ok
+    });
+
+    let result = block_on(async {
+        let a: DummyData = do_channel_read(123)
+            .await
+            .expect("Failed to read channel 123");
+        let b: DummyData = do_channel_read(456)
+            .await
+            .expect("Failed to read channel 456");
+
+        [a.0, b.0].join(", ")
+    })
+    .expect("block_on failed");
+
+    assert_eq!(result, "Hello, world!");
+}
+
+#[test]
+fn block_on_multiple_readers_parallel() {
+    init();
+    set_wait_on_channels_handler(|handles| {
+        for handle in handles.iter_mut() {
+            let data = match handle.handle() {
+                123 => "Hello",
```

Comment:
(optional) May make sense to declare these as consts within the test fn, e.g. `HANDLE_0` and `HANDLE_1`.

Reply:
Done!